src/tictactoe_program.rs

```diff
@@ -0,0 +1,434 @@
+impl std::error::Error for Error {}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum GridItem {
+    Free,
+    X,
+    O,
+}
+
+impl Default for GridItem {
+    fn default() -> GridItem {
+        GridItem::Free
+    }
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum State {
+    XMove,
+    OMove,
+    XWon,
+    OWon,
+    Draw,
+}
+impl Default for State {
+    fn default() -> State {
+        State::XMove
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize, PartialEq)]
+struct Game {
+    players: [Pubkey; 2],
+    state: State,
+    grid: [GridItem; 9],
+}
+
+impl Game {
+    pub fn new(player1: Pubkey, player2: Pubkey) -> Game {
+        let mut game = Game::default();
+        game.players = [player1, player2];
+        game
+    }
+
+    fn same(x_or_o: GridItem, triple: &[GridItem]) -> bool {
+        triple.iter().all(|&i| i == x_or_o)
+    }
+
+    pub fn next_move(self: &mut Game, player: Pubkey, x: usize, y: usize) -> Result<()> {
+        let index = self
+            .players
+            .iter()
+            .position(|&p| p == player)
+            .ok_or(Error::PlayerNotFound)?;
+
+        if self.grid[y * 3 + x] != GridItem::Free {
+            return Err(Error::InvalidMove);
+        }
+
+        let (x_or_o, won_state) = match self.state {
+            State::XMove => {
+                if index != 0 {
+                    return Err(Error::NotYourTurn)?;
+                }
+                self.state = State::OMove;
+                (GridItem::X, State::XWon)
+            }
+            State::OMove => {
+                if index != 1 {
+                    return Err(Error::NotYourTurn)?;
+                }
+                self.state = State::XMove;
+                (GridItem::O, State::OWon)
+            }
+            _ => {
+                return Err(Error::NotYourTurn)?;
+            }
+        };
+        self.grid[y * 3 + x] = x_or_o;
+
+        let winner =
+            // Check rows
+            Game::same(x_or_o, &self.grid[0..3])
+            || Game::same(x_or_o, &self.grid[3..6])
+            || Game::same(x_or_o, &self.grid[6..9])
+            // Check columns
+            || Game::same(x_or_o, &[self.grid[0], self.grid[3], self.grid[6]])
+            || Game::same(x_or_o, &[self.grid[1], self.grid[4], self.grid[7]])
+            || Game::same(x_or_o, &[self.grid[2], self.grid[5], self.grid[8]])
+            // Check both diagonals
+            || Game::same(x_or_o, &[self.grid[0], self.grid[4], self.grid[8]])
+            || Game::same(x_or_o, &[self.grid[2], self.grid[4], self.grid[6]]);
+
+        if winner {
+            self.state = won_state;
+        } else if self.grid.iter().all(|&p| p != GridItem::Free) {
+            self.state = State::Draw;
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+enum Command {
+    Init(Pubkey, Pubkey), // player1, player1
```

Comment:
player2

But why no command to add oneself to an existing game?

Reply:
In this setup player1 and player2 first rendezvous off chain.  player1 sends `Init()` with both their pub keys.   If player2 could add themselves to an existing game then I'd want to add an Accept command from player1 to prevent unwanted game crashers.